File: python_impl/preprocess.py
import os

import cv2
import numpy as np


# Диапазоны HSV подобраны под бледную красную сетку на почти белом фоне.
# Насыщенность снижена, потому что сетка слабая и малонасыщенная.
LOWER_RED_1 = np.array([0, 10, 120], dtype=np.uint8)
UPPER_RED_1 = np.array([20, 255, 255], dtype=np.uint8)
LOWER_RED_2 = np.array([160, 10, 120], dtype=np.uint8)
UPPER_RED_2 = np.array([180, 255, 255], dtype=np.uint8)

# Дополнительная эвристика по разности каналов.
DELTA_RG = 8
DELTA_RB = 8

# Порог по нормализованному каналу a из LAB.
A_THRESHOLD = 140
USE_LAB_RED_MASK = True

# Параметры морфологии для усиления тонкой сетки.
MORPH_SIZE = 3
DILATE_ITERATIONS = 2
OPEN_ITERATIONS = 1
HORIZONTAL_KERNEL_WIDTH = 25

# Параметры ROI поиска сетки.
SEARCH_Y_START = 0.08
SEARCH_Y_END = 0.95

# Параметры Canny и Hough.
CANNY_LOW = 50
CANNY_HIGH = 150
HOUGH_THRESHOLD = 35
HOUGH_MIN_LINE_FRAC = 0.15
HOUGH_MAX_LINE_GAP = 40

# Параметры фильтрации почти горизонтальных линий.
ANGLE_THRESHOLD = 10.0
MIN_FILTERED_LINES_WARNING = 5


def ensure_dir(path):
    """Создает директорию, если она отсутствует."""
    if path and not os.path.exists(path):
        os.makedirs(path, exist_ok=True)


def save_image(image_path, image):
    """Сохраняет изображение в файл."""
    ensure_dir(os.path.dirname(image_path))
    cv2.imwrite(image_path, image)


def load_image(image_path):
    """Загружает изображение с диска."""
    return cv2.imread(image_path)


def normalize_angle(angle_deg):
    """Приводит угол к диапазону около горизонтали."""
    if angle_deg > 90:
        angle_deg -= 180
    if angle_deg < -90:
        angle_deg += 180
    return angle_deg


def build_red_like_mask(image):
    """Строит маску красноватых пикселей по разности каналов."""
    blue, green, red = cv2.split(image)

    red_int = red.astype(np.int16)
    green_int = green.astype(np.int16)
    blue_int = blue.astype(np.int16)

    red_like = (red_int > green_int + DELTA_RG) & (red_int > blue_int + DELTA_RB)
    return (red_like.astype(np.uint8)) * 255


def build_lab_red_mask(image):
    """Строит дополнительную маску красноты по каналу a в LAB."""
    lab_image = cv2.cvtColor(image, cv2.COLOR_BGR2LAB)
    a_channel = lab_image[:, :, 1]
    lab_a_channel = cv2.normalize(a_channel, None, 0, 255, cv2.NORM_MINMAX)

    _, red_from_lab_mask = cv2.threshold(
        lab_a_channel,
        A_THRESHOLD,
        255,
        cv2.THRESH_BINARY,
    )

    return lab_a_channel, red_from_lab_mask


def extract_red_mask(image):
    """Выделяет бледную красную сетку несколькими способами и объединяет их."""
    hsv_image = cv2.cvtColor(image, cv2.COLOR_BGR2HSV)

    red_mask_1 = cv2.inRange(hsv_image, LOWER_RED_1, UPPER_RED_1)
    red_mask_2 = cv2.inRange(hsv_image, LOWER_RED_2, UPPER_RED_2)
    hsv_red_mask = cv2.bitwise_or(red_mask_1, red_mask_2)

    red_like_mask = build_red_like_mask(image)

    if USE_LAB_RED_MASK:
        lab_a_channel, red_from_lab_mask = build_lab_red_mask(image)
    else:
        lab_a_channel = np.zeros(image.shape[:2], dtype=np.uint8)
        red_from_lab_mask = np.zeros(image.shape[:2], dtype=np.uint8)

    red_mask_combined = cv2.bitwise_or(hsv_red_mask, red_like_mask)
    red_mask_combined = cv2.bitwise_or(red_mask_combined, red_from_lab_mask)

    kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (MORPH_SIZE, MORPH_SIZE))

    red_mask_after_close = cv2.morphologyEx(red_mask_combined, cv2.MORPH_CLOSE, kernel)
    red_mask_after_dilate = cv2.dilate(
        red_mask_after_close,
        kernel,
        iterations=DILATE_ITERATIONS,
    )
    red_mask_clean = cv2.morphologyEx(
        red_mask_after_dilate,
        cv2.MORPH_OPEN,
        kernel,
        iterations=OPEN_ITERATIONS,
    )

    debug_masks = {
        "red_mask_1": red_mask_1,
        "red_mask_2": red_mask_2,
        "red_like_mask": red_like_mask,
        "lab_a_channel": lab_a_channel,
        "red_from_lab_mask": red_from_lab_mask,
        "red_mask_combined_before_morph": red_mask_combined,
        "red_mask_after_close": red_mask_after_close,
        "red_mask_after_dilate": red_mask_after_dilate,
        "red_mask_clean": red_mask_clean,
    }

    return red_mask_clean, debug_masks


def build_search_roi_mask(mask):
    """Оставляет рабочую область без верхней служебной зоны и нижнего края."""
    height, width = mask.shape[:2]
    y_start = int(height * SEARCH_Y_START)
    y_end = int(height * SEARCH_Y_END)

    search_roi_mask = np.zeros((height, width), dtype=np.uint8)
    search_roi_mask[y_start:y_end, :] = mask[y_start:y_end, :]

    return search_roi_mask, y_start, y_end


def build_horizontal_emphasis_mask(mask):
    """Усиливает длинные горизонтальные структуры сетки."""
    kernel_horizontal = cv2.getStructuringElement(
        cv2.MORPH_RECT,
        (HORIZONTAL_KERNEL_WIDTH, 1),
    )
    return cv2.morphologyEx(mask, cv2.MORPH_OPEN, kernel_horizontal)


def run_hough(binary_image):
    """Запускает HoughLinesP по бинарному изображению."""
    _, width = binary_image.shape[:2]
    min_line_length = max(30, int(width * HOUGH_MIN_LINE_FRAC))

    raw_lines = cv2.HoughLinesP(
        binary_image,
        rho=1,
        theta=np.pi / 180,
        threshold=HOUGH_THRESHOLD,
        minLineLength=min_line_length,
        maxLineGap=HOUGH_MAX_LINE_GAP,
    )

    all_lines = []
    filtered_lines = []

    if raw_lines is None:
        return all_lines, filtered_lines

    for line in raw_lines:
        x1, y1, x2, y2 = line[0]
        angle = np.degrees(np.arctan2(y2 - y1, x2 - x1))
        angle = normalize_angle(float(angle))

        line_info = {
            "points": (x1, y1, x2, y2),
            "angle": angle,
        }
        all_lines.append(line_info)

        if abs(angle) < ANGLE_THRESHOLD:
            filtered_lines.append(line_info)

    return all_lines, filtered_lines


def detect_lines(red_mask_clean):
    """Ищет линии по маске сетки, а при необходимости переходит к fallback через edges."""
    search_roi_mask, roi_y_start, roi_y_end = build_search_roi_mask(red_mask_clean)
    horizontal_emphasis_mask = build_horizontal_emphasis_mask(search_roi_mask)

    if np.count_nonzero(horizontal_emphasis_mask) > 0:
        hough_input_mask = horizontal_emphasis_mask
        hough_input_name = "horizontal emphasis mask"
    else:
        hough_input_mask = search_roi_mask
        hough_input_name = "cleaned red mask ROI"

    mask_all_lines, mask_filtered_lines = run_hough(hough_input_mask)

    blurred_mask = cv2.GaussianBlur(search_roi_mask, (5, 5), 0)
    edges = cv2.Canny(blurred_mask, CANNY_LOW, CANNY_HIGH)
    edge_all_lines, edge_filtered_lines = run_hough(edges)

    selected_all_lines = mask_all_lines
    selected_filtered_lines = mask_filtered_lines
    hough_mode = "mask"
    selected_input_name = hough_input_name

    if len(mask_filtered_lines) < MIN_FILTERED_LINES_WARNING and len(edge_filtered_lines) > len(mask_filtered_lines):
        selected_all_lines = edge_all_lines
        selected_filtered_lines = edge_filtered_lines
        hough_mode = "edges fallback"
        selected_input_name = "edges"

    debug_data = {
        "search_roi_mask": search_roi_mask,
        "horizontal_emphasis_mask": horizontal_emphasis_mask,
        "hough_input_mask": hough_input_mask,
        "edges": edges,
        "roi_y_start": roi_y_start,
        "roi_y_end": roi_y_end,
        "hough_mode": hough_mode,
        "hough_input_name": selected_input_name,
        "mask_raw_lines": len(mask_all_lines),
        "mask_filtered_lines": len(mask_filtered_lines),
        "edge_raw_lines": len(edge_all_lines),
        "edge_filtered_lines": len(edge_filtered_lines),
    }

    return selected_all_lines, selected_filtered_lines, debug_data


def compute_rotation_angle(filtered_lines):
    """Вычисляет средний угол по почти горизонтальным линиям."""
    if not filtered_lines:
        return 0.0

    angle_sum = 0.0
    for line in filtered_lines:
        angle_sum += line["angle"]

    return angle_sum / len(filtered_lines)


def rotate_image(image, angle):
    """Поворачивает изображение вокруг центра."""
    height, width = image.shape[:2]
    center = (width // 2, height // 2)

    rotation_matrix = cv2.getRotationMatrix2D(center, angle, 1.0)

    return cv2.warpAffine(
        image,
        rotation_matrix,
        (width, height),
        flags=cv2.INTER_LINEAR,
        borderMode=cv2.BORDER_REPLICATE,
    )


def run_stage1_pipeline(image):
    """Выполняет этап 1 и возвращает ключевые результаты в удобном виде."""
    red_mask_clean, red_debug = extract_red_mask(image)
    all_lines, filtered_lines, line_debug = detect_lines(red_mask_clean)

    angle = compute_rotation_angle(filtered_lines)
    aligned_image = rotate_image(image, -angle)
    hough_lines_all = draw_lines(image, all_lines, (0, 255, 0))
    hough_lines_filtered = draw_lines(hough_lines_all, filtered_lines, (0, 0, 255))

    return {
        "original_image": image,
        "aligned_image": aligned_image,
        "final_angle": angle,
        "red_mask_clean": red_mask_clean,
        "red_debug": red_debug,
        "line_debug": line_debug,
        "all_lines": all_lines,
        "filtered_lines": filtered_lines,
        "hough_lines_all": hough_lines_all,
        "hough_lines_filtered": hough_lines_filtered,
    }


def draw_lines(image, lines, color):
    """Рисует линии поверх изображения."""
    debug_image = image.copy()

    for line in lines:
        x1, y1, x2, y2 = line["points"]
        cv2.line(debug_image, (x1, y1), (x2, y2), color, 2)

    return debug_image


def prepare_collage_tile(image, width, height, title):
    """Подготавливает одно изображение для debug-коллажа."""
    if len(image.shape) == 2:
        tile = cv2.cvtColor(image, cv2.COLOR_GRAY2BGR)
    else:
        tile = image.copy()

    tile = cv2.resize(tile, (width, height), interpolation=cv2.INTER_AREA)
    cv2.putText(
        tile,
        title,
        (10, 24),
        cv2.FONT_HERSHEY_SIMPLEX,
        0.7,
        (0, 255, 255),
        2,
        cv2.LINE_AA,
    )
    return tile


def create_debug_collage(images):
    """Собирает коллаж 3x3 из ключевых debug-изображений."""
    cell_width = 420
    cell_height = 260
    rows = []

    for row_index in range(0, len(images), 3):
        row_images = images[row_index:row_index + 3]
        tiles = []

        for title, image in row_images:
            tiles.append(prepare_collage_tile(image, cell_width, cell_height, title))

        while len(tiles) < 3:
            tiles.append(np.zeros((cell_height, cell_width, 3), dtype=np.uint8))

        rows.append(np.hstack(tiles))

    return np.vstack(rows)

File: cpp_impl/src/preprocess.cpp
#include "preprocess.hpp"

#include <algorithm>
#include <cmath>
#include <filesystem>
#include <iostream>
#include <numeric>
#include <string>
#include <vector>

namespace {

constexpr int DELTA_RG = 8;
constexpr int DELTA_RB = 8;
constexpr int A_THRESHOLD = 140;
constexpr int MORPH_SIZE = 3;
constexpr int DILATE_ITERATIONS = 2;
constexpr int OPEN_ITERATIONS = 1;
constexpr int HORIZONTAL_KERNEL_WIDTH = 25;
constexpr double SEARCH_Y_START = 0.08;
constexpr double SEARCH_Y_END = 0.95;
constexpr int CANNY_LOW = 50;
constexpr int CANNY_HIGH = 150;
constexpr int HOUGH_THRESHOLD = 30;
constexpr double HOUGH_MIN_LINE_FRAC = 0.15;
constexpr int HOUGH_MAX_LINE_GAP = 40;
constexpr double ANGLE_THRESHOLD = 15.0;
constexpr int MIN_FILTERED_LINES_WARNING = 5;

std::vector<cv::Vec4i> g_lastRawLines;
std::vector<cv::Vec4i> g_lastFilteredLines;

void saveImage(const std::string& path, const cv::Mat& image) {
    ensureDir(std::filesystem::path(path).parent_path().string());
    cv::imwrite(path, image);
}

int countPixels(const cv::Mat& mask) {
    return cv::countNonZero(mask);
}

void printMaskPixels(const std::string& name, const cv::Mat& mask) {
    std::cout << "[INFO] " << name << " pixels: " << countPixels(mask) << "\n";
}

cv::Mat buildRedLikeMask(const cv::Mat& image) {
    std::vector<cv::Mat> channels;
    cv::split(image, channels);

    cv::Mat blue16;
    cv::Mat green16;
    cv::Mat red16;
    channels[0].convertTo(blue16, CV_16S);
    channels[1].convertTo(green16, CV_16S);
    channels[2].convertTo(red16, CV_16S);

    cv::Mat redGtGreen;
    cv::Mat redGtBlue;
    cv::compare(red16, green16 + DELTA_RG, redGtGreen, cv::CMP_GT);
    cv::compare(red16, blue16 + DELTA_RB, redGtBlue, cv::CMP_GT);

    cv::Mat redLikeMask;
    cv::bitwise_and(redGtGreen, redGtBlue, redLikeMask);
    return redLikeMask;
}

void buildLabRedMask(const cv::Mat& image, cv::Mat& labAChannel, cv::Mat& redFromLabMask) {
    cv::Mat labImage;
    cv::cvtColor(image, labImage, cv::COLOR_BGR2Lab);

    std::vector<cv::Mat> labChannels;
    cv::split(labImage, labChannels);
    cv::normalize(labChannels[1], labAChannel, 0, 255, cv::NORM_MINMAX);
    cv::threshold(labAChannel, redFromLabMask, A_THRESHOLD, 255, cv::THRESH_BINARY);
}

double normalizeLineAngle(double angleDeg) {
    // Приведение к диапазону около горизонтали по правилам ТЗ.
    if (angleDeg > 45.0) {
        angleDeg -= 90.0;
    }
    if (angleDeg < -45.0) {
        angleDeg += 90.0;
    }
    return angleDeg;
}

double lineAngle(const cv::Vec4i& line) {
    const double dx = static_cast<double>(line[2] - line[0]);
    const double dy = static_cast<double>(line[3] - line[1]);
    return normalizeLineAngle(std::atan2(dy, dx) * 180.0 / CV_PI);
}

double lineLength(const cv::Vec4i& line) {
    const double dx = static_cast<double>(line[2] - line[0]);
    const double dy = static_cast<double>(line[3] - line[1]);
    return std::sqrt(dx * dx + dy * dy);
}

std::vector<cv::Vec4i> runHough(const cv::Mat& binaryImage) {
    std::vector<cv::Vec4i> lines;
    const int minLineLength = std::max(30, static_cast<int>(binaryImage.cols * HOUGH_MIN_LINE_FRAC));
    cv::HoughLinesP(
        binaryImage,
        lines,
        1,
        CV_PI / 180.0,
        HOUGH_THRESHOLD,
        minLineLength,
        HOUGH_MAX_LINE_GAP
    );
    return lines;
}

std::vector<cv::Vec4i> filterHorizontalLines(const std::vector<cv::Vec4i>& lines, int imageWidth) {
    std::vector<cv::Vec4i> filtered;
    const double minLength = imageWidth * HOUGH_MIN_LINE_FRAC;

    for (const cv::Vec4i& line : lines) {
        const double angle = lineAngle(line);
        const double length = lineLength(line);

        if (std::abs(angle) < ANGLE_THRESHOLD && length > minLength) {
            filtered.push_back(line);
        }
    }

    return filtered;
}

double median(std::vector<double> values) {
    if (values.empty()) {
        return 0.0;
    }

    std::sort(values.begin(), values.end());
    const size_t middle = values.size() / 2;
    if (values.size() % 2 == 1) {
        return values[middle];
    }
    return (values[middle - 1] + values[middle]) / 2.0;
}

}  // namespace

void ensureDir(const std::string& path) {
    if (!path.empty()) {
        std::filesystem::create_directories(path);
    }
}

cv::Mat loadImage(const std::string& jpgPath, const std::string& pngPath) {
    if (std::filesystem::exists(jpgPath)) {
        cv::Mat image = cv::imread(jpgPath, cv::IMREAD_COLOR);
        if (!image.empty()) {
            return image;
        }
    }

    std::cout << "[WARN] Файл " << jpgPath << " не найден, используется " << pngPath << "\n";
    return cv::imread(pngPath, cv::IMREAD_COLOR);
}

cv::Mat extractRedMask(const cv::Mat& image, const std::string& debugDir) {
    cv::Mat hsvImage;
    cv::cvtColor(image, hsvImage, cv::COLOR_BGR2HSV);

    cv::Mat redMask1;
    cv::Mat redMask2;
    cv::inRange(hsvImage, cv::Scalar(0, 10, 120), cv::Scalar(20, 255, 255), redMask1);
    cv::inRange(hsvImage, cv::Scalar(160, 10, 120), cv::Scalar(180, 255, 255), redMask2);

    cv::Mat hsvRedMask;
    cv::bitwise_or(redMask1, redMask2, hsvRedMask);

    cv::Mat redLikeMask = buildRedLikeMask(image);

    cv::Mat labAChannel;
    cv::Mat redFromLabMask;
    buildLabRedMask(image, labAChannel, redFromLabMask);

    cv::Mat redMaskCombined;
    cv::bitwise_or(hsvRedMask, redLikeMask, redMaskCombined);
    cv::bitwise_or(redMaskCombined, redFromLabMask, redMaskCombined);

    cv::Mat kernel = cv::getStructuringElement(cv::MORPH_RECT, cv::Size(MORPH_SIZE, MORPH_SIZE));

    cv::Mat redMaskAfterClose;
    cv::Mat redMaskAfterDilate;
    cv::Mat redMaskClean;
    cv::morphologyEx(redMaskCombined, redMaskAfterClose, cv::MORPH_CLOSE, kernel);
    cv::dilate(redMaskAfterClose, redMaskAfterDilate, kernel, cv::Point(-1, -1), DILATE_ITERATIONS);
    cv::morphologyEx(
        redMaskAfterDilate,
        redMaskClean,
        cv::MORPH_OPEN,
        kernel,
        cv::Point(-1, -1),
        OPEN_ITERATIONS
    );

    saveImage(debugDir + "/red_mask_1.png", redMask1);
    saveImage(debugDir + "/red_mask_2.png", redMask2);
    saveImage(debugDir + "/red_like_mask.png", redLikeMask);
    saveImage(debugDir + "/lab_a_channel.png", labAChannel);
    saveImage(debugDir + "/red_from_lab_mask.png", redFromLabMask);
    saveImage(debugDir + "/red_mask_combined_before_morph.png", redMaskCombined);
    saveImage(debugDir + "/red_mask_after_close.png", redMaskAfterClose);
    saveImage(debugDir + "/red_mask_after_dilate.png", redMaskAfterDilate);
    saveImage(debugDir + "/red_mask_clean.png", redMaskClean);

    printMaskPixels("red_mask_1", redMask1);
    printMaskPixels("red_mask_2", redMask2);
    printMaskPixels("red_like_mask", redLikeMask);
    printMaskPixels("lab_red_mask", redFromLabMask);
    printMaskPixels("red_mask_clean", redMaskClean);

    return redMaskClean;
}

std::vector<cv::Vec4i> detectHorizontalLines(
    const cv::Mat& mask,
    const cv::Size& fullSize,
    const std::string& debugDir
) {
    const int yStart = static_cast<int>(fullSize.height * SEARCH_Y_START);
    const int yEnd = static_cast<int>(fullSize.height * SEARCH_Y_END);

    cv::Mat searchRoiMask = cv::Mat::zeros(mask.size(), mask.type());
    mask.rowRange(yStart, yEnd).copyTo(searchRoiMask.rowRange(yStart, yEnd));
    saveImage(debugDir + "/search_roi_mask.png", searchRoiMask);

    const int roiHeight = yEnd - yStart;
    std::cout << "[INFO] Search ROI: y=" << yStart << ":" << yEnd
              << " (" << fullSize.width << "x" << roiHeight << ")\n";

    cv::Mat kernelHorizontal = cv::getStructuringElement(
        cv::MORPH_RECT,
        cv::Size(HORIZONTAL_KERNEL_WIDTH, 1)
    );
    cv::Mat horizontalEmphasisMask;
    cv::morphologyEx(searchRoiMask, horizontalEmphasisMask, cv::MORPH_OPEN, kernelHorizontal);
    saveImage(debugDir + "/horizontal_emphasis_mask.png", horizontalEmphasisMask);
    saveImage(debugDir + "/hough_input_mask.png", horizontalEmphasisMask);

    printMaskPixels("horizontal emphasis mask", horizontalEmphasisMask);

    std::vector<cv::Vec4i> maskRawLines = runHough(horizontalEmphasisMask);
    std::vector<cv::Vec4i> maskFilteredLines = filterHorizontalLines(maskRawLines, fullSize.width);

    cv::Mat blurredMask;
    cv::Mat edges;
    cv::GaussianBlur(searchRoiMask, blurredMask, cv::Size(5, 5), 0);
    cv::Canny(blurredMask, edges, CANNY_LOW, CANNY_HIGH);
    saveImage(debugDir + "/edges.png", edges);

    g_lastRawLines = maskRawLines;
    g_lastFilteredLines = maskFilteredLines;

    if (static_cast<int>(maskFilteredLines.size()) < MIN_FILTERED_LINES_WARNING) {
        std::cout << "[WARN] Hough по mask не дал линий, используется fallback edges\n";
        g_lastRawLines = runHough(edges);
        g_lastFilteredLines = filterHorizontalLines(g_lastRawLines, fullSize.width);
    }

    std::cout << "[INFO] Hough raw lines: " << g_lastRawLines.size() << "\n";
    std::cout << "[INFO] Hough filtered horizontal lines: " << g_lastFilteredLines.size() << "\n";

    return g_lastFilteredLines;
}

double computeRotationAngle(const std::vector<cv::Vec4i>& lines) {
    if (lines.empty()) {
        std::cout << "[WARN] Горизонтальные линии не найдены, используется угол 0 градусов\n";
        return 0.0;
    }

    std::vector<double> angles;
    angles.reserve(lines.size());
    for (const cv::Vec4i& line : lines) {
        angles.push_back(lineAngle(line));
    }

    const double firstMedian = median(angles);
    std::vector<double> trimmedAngles;
    for (double angle : angles) {
        if (angle >= firstMedian - 5.0 && angle <= firstMedian + 5.0) {
            trimmedAngles.push_back(angle);
        }
    }

    if (trimmedAngles.empty()) {
        return firstMedian;
    }

    return median(trimmedAngles);
}

cv::Mat rotateImage(const cv::Mat& image, double angleDeg) {
    const cv::Point2f center(
        static_cast<float>(image.cols) / 2.0f,
        static_cast<float>(image.rows) / 2.0f
    );
    cv::Mat rotationMatrix = cv::getRotationMatrix2D(center, angleDeg, 1.0);

    cv::Mat rotated;
    cv::warpAffine(
        image,
        rotated,
        rotationMatrix,
        image.size(),
        cv::INTER_LINEAR,
        cv::BORDER_REPLICATE
    );
    return rotated;
}

void drawLines(
    const cv::Mat& image,
    const std::vector<cv::Vec4i>& lines,
    const std::string& outputPath,
    const cv::Scalar& color
) {
    cv::Mat debugImage;
    if (image.channels() == 1) {
        cv::cvtColor(image, debugImage, cv::COLOR_GRAY2BGR);
    } else {
        debugImage = image.clone();
    }

    for (const cv::Vec4i& line : lines) {
        cv::line(
            debugImage,
            cv::Point(line[0], line[1]),
            cv::Point(line[2], line[3]),
            color,
            2,
            cv::LINE_AA
        );
    }

    saveImage(outputPath, debugImage);
}

const std::vector<cv::Vec4i>& getLastRawLines() {
    return g_lastRawLines;
}

int getLastRawLineCount() {
    return static_cast<int>(g_lastRawLines.size());
}

int getLastFilteredLineCount() {
    return static_cast<int>(g_lastFilteredLines.size());
}
